sound/soc/sof/intel/hda-dai.c

```diff
@@ -475,17 +480,78 @@ static int ssp_dai_hw_params(struct snd_pcm_substream *substream,
 			     struct snd_pcm_hw_params *params,
 			     struct snd_soc_dai *dai)
 {
+	struct ssp_dai_dma_data *dma_data;
+
+	dma_data = kzalloc(sizeof(*dma_data), GFP_KERNEL);
```

Comment:
Check if we already have the dma_data, hw_params can be called several times.
It might be wise to add `.startup. and allocate the dma_data there. It would ensure that we have DMA data as long as the `.hw_free` is called.

Reply:
valid point, we should add a startup and shutdown case, that was the alloc/free is safe. Good suggestion indeed.

Reply:
I did the move, but now I have a doubt. @ranj063 @ujfalusi what happens if we do the hw_params multiple times? Are we going to setup the widgets and send an IPC multiple times?

Now that I think of it, I think there can be cases where hw_free can be called when hw_params failed on the FE but was not setup in the BE, we had such an issue with the HDA link.

Maybe we need to harden the hw_params/hw_free as well, in addition to the changes made so far for suspend/resume?

Reply:
see suggested update @ranj063 @ujfalusi , I added a separate flag to setup the widgets once in hw_params and only free them in hw_free if they've actually been allocated.

Reply:
I don't know if it can ever happen, but we can end up in a non configured DAI if after the `.trigger:SNDRV_PCM_TRIGGER_SUSPEND` the `.hw_params` is called first instead of the `.prepare`.

configured is true and in hw_params we set the suspended to false so the `ssp_dai_setup_or_free(substream, dai, true)` will be not called from hw_params nor from the prepare.

Reply:
@plbossart hw_params can be called multiple times only only either after a stop trigger or a suspend both followed by a prepare. In the latter case (for stop), you've already eliminated hw_params because of the suspended flag. But hw_free can still happen without restarting the stream after resume and we handle the case with the configured flag in snd_sof_dai

Reply:
@ranj063 I am not a big fan of this sof_dai::configured flag. We're blurring layers between dais and widgets, if we introduce a flag it should deal with the different callbacks at the same level. I have a much simpler proposal that I will push now.